src/UI/Implementation/Factory.php

```diff
@@ -4,41 +4,175 @@
 
 namespace ILIAS\UI\Implementation;
 
+use \ILIAS\UI\Component as C;
 use ILIAS\UI\NotImplementedException;
 
 // TODO: This might cache the created factories.
 use ILIAS\UI\Implementation\Component\SignalGenerator;
 
 class Factory implements \ILIAS\UI\Factory
 {
+	/**
+	 * @var C\Counter\Factory
+	 */
+	protected $counter_factory;
+
+	/**
+	 * @var C\Glyph\Factory
+	 */
+	protected $glyph_factory;
+
+	/**
+	 * @var C\Button\Factory
+	 */
+	protected $button_factory;
+
+	/**
+	 * @var C\Listing\Factory
+	 */
+	protected $listing_factory;
+
+	/**
+	 * @var C\Image\Factory
+	 */
+	protected $image_factory;
+
+	/**
+	 * @var C\Panel\Factory
+	 */
+	protected $panel_factory;
+
+	/**
+	 * @var C\Modal\Factory
+	 */
+	protected $modal_factory;
+
+	/**
+	 * @var C\Dropzone\Factory
+	 */
+	protected $dropzone_factory;
+
+	/**
+	 * @var C\Popover\Factory
+	 */
+	protected $popover_factory;
+
+	/**
+	 * @var C\Divider\Factory
+	 */
+	protected $divider_factory;
+
+	/**
+	 * @var C\Link\Factory
+	 */
+	protected $link_factory;
+
+	/**
+	 * @var C\Dropdown\Factory
+	 */
+	protected $dropdown_factory;
+
+	/**
+	 * @var C\Item\Factory
+	 */
+	protected $item_factory;
+
+	/**
+	 * @var C\Icon\Factory
+	 */
+	protected $icon_factory;
+
+	/**
+	 * @var C\ViewControl\Factory
+	 */
+	protected $viewcontrol_factory;
+
+	/**
+	 * @var C\Chart\Factory
+	 */
+	protected $chart_factory;
+
+	/**
+	 * @var C\Input\Factory
+	 */
+	protected $input_factory;
+
+	/**
+	 * @var C\Table\Factory
+	 */
+	protected $table_factory;
+
+	/**
+	 * @var Component\MessageBox\Factory
+	 */
+	protected $messagebox_factory;
+
+	public function __construct(
+		C\Counter\Factory $counter_factory,
+		C\Glyph\Factory $glyph_factory,
+		C\Button\Factory $button_factory,
+		C\Listing\Factory $listing_factory,
+		C\Image\Factory	$image_factory,
+		C\Panel\Factory $panel_factory,
+		C\Modal\Factory $modal_factory,
+		C\Dropzone\Factory $dropzone_factory,
+		C\Popover\Factory $popover_factory,
+		C\Divider\Factory $divider_factory,
+		C\Link\Factory $link_factory,
+		C\Dropdown\Factory $dropdown_factory,
+		C\Item\Factory $item_factory,
+		C\Icon\Factory $icon_factory,
+		C\ViewControl\Factory $viewcontrol_factory,
+		C\Chart\Factory $chart_factory,
+		C\Input\Factory $input_factory,
+		C\Table\Factory $table_factory,
+		C\MessageBox\Factory $messagebox_factory
+	) {
```

Comment:
What about using a optional dependency data object that may contain these factories. If factories are present in the dependeny data object they are type checked and used, if not we fall back to call "new ...". This way the signature won't blow up and single changes to the default factories can be injected easily. In the future we may hopefully use https://wiki.php.net/rfc/typed_properties_v2 to do proper typisation of these objects.

Reply:
I do not see a reason to hide the dependencies this object really has. "Signature blow up" might be a problem for some method, I cannot see why this would be a problem here. In the end one never needs to instantiate the object in productive code.

The long list of deps rather is a symptom of another problem we have here: the top level of the component tree gets kind of messy an would benefit from some additional structure. This is also the strategy we should use to solve similar issues (many deps) in other components: break the component in smaller parts with less dependencies.

Reply:
I think maintainability is something you want also in your test code. Now the signature is spread already in

tests/UI/Component/Breadcrumbs/BreadcrumbsTest.php
tests/UI/Component/Card/CardTest.php
tests/UI/Component/Deck/DeckTest.php
tests/UI/Component/Dropzone/File/DropzoneRendererTest.php
tests/UI/Component/Legacy/LegacyTest.php
tests/UI/Component/MessageBox/MessageBoxTest.php
tests/UI/Component/Modal/ModalBase.php
tests/UI/Component/ViewControl/PaginationTest.php
tests/UI/Component/ViewControl/SortationTest.php
tests/UI/Component/ViewControl/ViewControlTest.php

All these test classes need to be rewritten each time a new top level node is introduced. And most of the test do not really depend on the other components. The factory does. A MockFactory or some other helper class could hide all this.

Reply:
The maintainability is indeed better if dependencies are made explicit, because devs can then see the deps and PHP can warn you that something is missing very early in the process. I see your point about the spread of the signature to the tests, though, and agree on hiding that in some helper.